### PR TITLE
Upgrade to mongo driver 4.x and java 8

### PR DESCRIPTION
**Pull request by Dayde**
Driver breaking changes :
* JSON class was removed, use BasicDBObject instead
* ObjectId.createFromLegacyFormat method was removed use new ObjectId() instead
* no longer java 1.5 compatible hence the upgrade to java 8
* $maxScan operator no longer supported
* BsonCodec does not implement the decode method
* UUID are no longer serialized using $uuid operator
* null countOption no longer supported

Also upgrade
* de.flapdoodle.embed.mongo 2.2.0 -> 3.0.0
* de.undercouch.bson4jackson 2.9.2 -> 2.12.0

### CODE REVIEW

**Comment by Dayde** [comment]
Hi, as Jongo users at Malt we wanted to be able to upgrade to mongo diver 4.x.

There’s no rush so we thought it would be nice to have it in the official repository, hence this pull request :)

Looking forward to read your reviews.

**Comment by bguerout** [comment]
Hello,

First of all thank you very much for taking the time to do this pull request.
At first glance it seems ok to me. I will do a review and possibly put some comments if needed.

**Comment by Daemorinumi** [comment]
Hello,

Do you know when the release will be available ?

Have a nice day

**Comment by ndemengel** [comment]
Hi Benoît,

Long time no see :-)

Can we help you in any way to decide on this PR? Not that we want to rush you, we know perfectly well what it costs to maintain OSS, and even more when one doesn't actively develop on it anymore.
I've just thought it owuld be a good idea to ask, just in case :-)

Edit: funnily enough, the previous comment appeared while I was typing mine. We didn't consult each other (and don't even know each other), promise!

**Review by bguerout — COMMENTED** [review]
Sorry for the delay, it's not nice considering the effort involved.
Here is few comments.

File: pom.xml [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
     <groupId>org.jongo</groupId>
     <artifactId>jongo</artifactId>
-    <version>1.5.0-SNAPSHOT</version>
+    <version>2.0.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (Dayde) [reply pom.xml:24]: The drop of java < 8 support was what motivated bumping the major version, but if you tell me minor is enough, minor it is :)

File: src/main/java/org/jongo/JongoNative.java [hunk 2]
```diff
@@ -54,9 +55,9 @@ public JongoNative(MongoDatabase database, Mapper mapper) {
         return wrapCollection(collection);
     }
 
-    public Bson query(String query, Object... parameters) {
+    public BsonDocument query(String query, Object... parameters) {
```
[the remark this answers is not in the document]
Comment (Dayde) [reply src/main/java/org/jongo/JongoNative.java:58]: No I’m not using it, I tried my best no to break anything or change contracts but this was kind of hard around this as the `Query.toBson` method has been removed in mongo 4.

**Comment by Dayde** [comment]
No worry about delays we all have our jobs and lives :)
The time we took to make this pull request is nothing compared to the time you took to build the lib itself and maintain it.
Plus it was a pleasure to work on such a nice code base ! Tests were really helpful during this upgrade.

**Comment by Daemorinumi** [comment]
Hello,

Maybe i'm wrong but I think that upgrade is not complete.

There are for example a declaration of DB and DBCollection in the class /src/main/java/org/jongo/Jongo.java
DB and DBCollection are not longer available for the last mongo version (ie mongodb-driver-sync).

It's possible to use mongodb-driver-legacy which include these classe, but it's a legacy.

source : https://mongodb.github.io/mongo-java-driver/4.2/driver/getting-started/installation/

Have a nice day

**Comment by bguerout** [comment]
@Dayde Hello, i've deployed a snapshot version, are you able to check if this snapshot works with your codebase ?

```
  <dependency>
    <groupId>org.jongo</groupId>
    <artifactId>jongo</artifactId>
    <version>1.5.0-SNAPSHOT</version>
  </dependency>
```

@Daemorinumi This PR allows Jongo to be able to still rely on legacy API (ie.`DBCollection`) using `mongodb-driver-legacy`.
This legacy driver contains also the "new" API (ie. `MongoCollection`). You can have a look at #381 which tries to deal with `MongoCollection`

**Comment by Daemorinumi** [comment]
Hello @bguerout,

Thank you

**Comment by Dayde** [comment]
Hi, sure, I’ll test that as soon as possible (should be done by the end of next week).

**Comment by Dayde** [comment]
Hi, it took a bit longer than planned but here we are :) the upgrade is under test and we spotted some regressions.
Here are some fixes #385

**Comment by jmesny** [comment]
Thanks @Dayde for the contribution and thanks @bguerout for the review.
On our side we tested this `1.5.0-SNAPSHOT` as dependency for our codebase that heavily relies on Jongo.
We didn't detect any issues, everything works fine!
@bguerout Do you plan to release it as a stable version soon or do you want to also include in this `1.5.0` the work you started in #381?
We're very happy to see Jongo evolve! Thanks a lot! 😃

**Comment by Dayde** [comment]
Thanks for testing @jmesny !

We did some testing too and found that the upgrade fixed some weird behaviour of the parser like accepting trailing curly braces.

Maybe something to advertise in the changelog for the release.

Following test passes with latest version but not in `1.4` as no exception is thrown.
```
@Test(expected = IllegalArgumentException.class)
public void shouldNotAllowTrailingCurlyBraces() throws Exception {
    Query query = factory.createQuery("{id:'123'}}");
    assertThat(query.toDBObject()).isEqualTo(QueryBuilder.start("id").is("123").get());
}
```

**Comment by bguerout** [comment]
Hello,
Thanks for your feedbacks.
I will release 1.5.0 tomorrow.

@jmesny #381 will not be included

**Comment by bguerout** [comment]
Hi,

Jongo 1.5.0 has been released.
I will update the website http://jongo.org/

Thanks again for the great work.

**Pull request merged**